autopush-common/src/db/bigtable/pool.rs

```diff
@@ -0,0 +1,193 @@
+                .ok_or_else(|| DbError::ConnectionError(format!(
+                    "Invalid DSN: Unparsable host {:?}",
+                    endpoint
+                )))?,
+            parsed.port().unwrap_or(8086)
+        );
+        if !parsed.path().is_empty() {
+            return Err(DbError::ConnectionError(format!(
+                "Invalid DSN: Table paths belong in settings : {:?}",
+                endpoint
+            )));
+        }
+        let connection = format!("{}{}", origin, parsed.path());
+        debug!("🉑 connection string {}", &connection);
+
+        // Construct a new manager and put them in a pool for handling future requests.
+        let manager =
+            BigtableClientManager::new(settings, settings.dsn.clone(), connection.clone())?;
+        let mut config = PoolConfig::default();
+        if let Some(size) = bt_settings.database_pool_max_size {
+            debug!("🏊 Setting pool max size {}", &size);
+            config.max_size = size as usize;
+        };
+        if let Some(timeout) = bt_settings.database_pool_connection_timeout {
+            debug!("🏊 Setting connection timeout to {} milliseconds", &timeout);
+            let timeouts = Timeouts {
+                create: Some(Duration::from_millis(timeout as u64)),
+                ..Default::default()
+            };
+            config.timeouts = timeouts;
+        }
+        let pool = deadpool::managed::Pool::builder(manager)
+            .config(config)
+            .build()
+            .map_err(|e| DbError::BTError(BigTableError::Pool(e.to_string())))?;
+
+        Ok(Self { pool })
+    }
+}
+
+/// BigTable Pool Manager. This contains everything needed to create a new connection.
+pub struct BigtableClientManager {
+    settings: DbSettings,
+    dsn: Option<String>,
+    connection: String,
+}
+
+impl BigtableClientManager {
+    fn new(
+        settings: &DbSettings,
+        dsn: Option<String>,
+        connection: String,
+    ) -> Result<Self, DbError> {
+        Ok(Self {
+            settings: settings.clone(),
+            dsn,
+            connection,
+        })
+    }
+}
+
+impl fmt::Debug for BigtableClientManager {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("deadpool::BtClientManager")
+            .field("settings", &self.settings.clone())
+            .finish()
+    }
+}
+
+#[async_trait]
+impl Manager for BigtableClientManager {
+    type Error = DbError;
+    type Type = BigtableClient;
+
+    /// Create a new Bigtable Client with it's own channel.
+    /// `BigtableClient` is the most atomic we can go.
+    async fn create(&self) -> Result<BigtableClient, DbError> {
+        debug!("🏊 Create a new pool entry.");
+        let chan = Self::create_channel(self.dsn.clone())?.connect(self.connection.as_str());
+        let client = BigtableClient::new(chan);
+        Ok(client)
+    }
+
+    /// We can't really recycle a given client, so fail and the client should be dropped.
```

Comment:
It's a bit complicated. The actual connection socket element is buried somewhat inside of the client, and it's private so we can't just re-establish the connection. There's also potentially some issues with state that could be preserved, particularly in future versions, so I don't want to crack open something or use it in anything other than the dumbest possible way.

Reply:
The connection doesn't necessarily need to be re-established during deadpool's recycle. It's a lifetime event occuring when [pulling an existing Object out of its pool](https://github.com/bikeshedder/deadpool/blob/master/src/managed/mod.rs#L355).

So if anything recycle should probably be validating the connection.

In syncstorage we do that and also use this as a hook for the lifetime configuration of the connection (e.g. drop it after 60 minutes of usage). There should be an [alternative way](https://github.com/mozilla-services/syncstorage-rs/issues/1419) of doing that lifetime management via a background task that we could take advantage of in the later version of deadpool we're using here, instead of how syncstorage does this.

But our recycle here could still do a health check of the connection. In `diesel`'s similar hook for the `r2d2` pool this it's usually something like ensuring a `SELECT 1` query succeeded. For spanner we call `get_session` instead.

For some reason I can't recall, in syncstorage we also re-establish connections in recycle when they're disconnected. I don't think that's entirely necessary: instead returning a `RecycleError` should signal to the pool to drop that Object and silently move on to trying a different Object from the pool or creating a new one.

Reply:
Huh, BigtableClient doesn't really expose any traits for doing a general connection check, but we could probably fake one by attempting to do a read. Let me look into what can be done.

Reply:
I added some TTLs for connection and idle checks for now, and pulled the `health_check` lower so that we can use it for the pool check.